### DISCUSSION TITLE
8 bpc values with Q16 build?

### DISCUSSION THREAD

**Discussion by blue-j** [ImageMagick/ImageMagick#6576] [Help] [answered]
### ImageMagick version

7.1.1-14 Q16-HDRI aarch64

### Operating system, version and so on

macOS 13.4.1 (c)

### Description

i have images that are 8bpc.

I run

`magick identify -format '%[min] %[max] %[mean] %[median] %[standard-deviation] %[kurtosis] %[skewness] ' filename.ext`

and the stats print out in 16 bpc, since my build is Q16.  How can I make the output 8 bpc values?

- J

**Comment by snibgo** [comment]
`%[min]` etc are on a nominal scale of 0 to QuantumLevel, which is 255 for Q8, 65535 for Q16, etc.

By contrast, `%[fx:minima]` etc are on a nominal scale of 0.0 to 1.0 for all Q-values. In addition, they can take expressions. So you can use `%[fx:minima*255]` to get a scale of 0 to 255.

**Comment by blue-j** [comment]
`magick identify -format '%[fx:minima*255] %[fx:maxima*255] %[fx:mean*255] %[fx:median*255] %[fx:standard_deviation*255] %[fx:kurtosis*255] %[fx:skewness*255] filename.ext`

does this look good?  AFAICT, the output is correct, but i am feeling a little uneasy about kurtosis and skewness, which are pivotal to my analysis.

you are a god!

**Reply by fmw42** [reply]
Kurtosis nominally is between -3 and 3.  See https://en.wikipedia.org/wiki/Kurtosis#:~:text=The%20kurtosis%20is%20bounded%20below,and%20it%20may%20be%20infinite.

Skewness nominally between -1 and 1, but can exceed that range. -2 to 2 is highly skewed.  See https://www.smartpls.com/documentation/functionalities/excess-kurtosis-and-skewness/

**Comment by snibgo** [comment, marked as answer]
@blue-j: Yes, that is fine, except for ...

> but i am feeling a little uneasy about kurtosis and skewnes

You and me both. I don't know the scale of these. The `fx.c` module multiplies these value by QuantumScale (eg 1/255, 1/65535), but I'm not sure if it should. It always did multiply by QuantumScale, so that is what the "new" (for a couple of years) fx does.


In statistic.c, GetImageStatistics() calculates:
```
skewness = ( sum_cubed - 3*mean*sum_squared + 2*mean*mean*mean ) * 1/(SD*SD*SD)

kurtosis = ( sum_fourth_power
             - 4.0*mean*sum_cubed
             + 6.0*mean*mean*sum_squared
             - 3.0*mean*mean*1.0*mean*mean
           ) * 1/(SD*SD*SD*SD)
           - 3.0;
```
If we change QuantumLevel (the maximum value of pixels) from 255 to 65535, each component (sum_cubed, SD, etc) is multiplied by the same amount, so I think skewness and kurtosis have values that are independent of QuantumLevel, and hence of the Q-number.

We can confirm this:
```
D:\temp>%IMG7%magick identify  -format "%[kurtosis] %[skewness]\n" toes.png
0.0394962 0.364843

D:\temp>%IMG7%magick identify  -format "%[fx:kurtosis*QuantumRange] %[fx:skewness*QuantumRange]\n" toes.png
0.0394962 0.364843
```

I conclude that if we use kurtosis or skewness within fx, we should multiply them by QuantumRange, to undo the effect of multiplying by QuantumScale. Or just use "`%[kurtosis] %[skewness]`" without fx. But I'm not an expert at this, and happy to be corrected.

[While writing this, fmw42 replied. Thanks, fmw42. I think that is consistent with what I have written.]

**Comment by blue-j** [comment]
More testing has produced vexation:

-0.0956531    -format %[kurtosis]
-0.00000145957    -format %[fx:kurtosis]
-0.000372191    -format %[fx:kurtosis*255]
-0.0786    manual calculation (numpy in python)       kurtosis = np.average(((values - mean)/std)**4, weights=bin) - 3

sigh.  not sure which is correct!
![test](https://github.com/ImageMagick/ImageMagick/assets/3990142/2beb9b17-66d1-4366-8f88-0e830290004b)

- J

**Comment by blue-j** [comment]
I believe our manual calculation (written by someone with a PhD in statistics, not me) is "excess kurtosis"?   - J

**Comment by blue-j** [comment]
-0.342 from ImageJ

i think i may start drinking

![Screenshot 2023-08-21 at 9 08 02 PM](https://github.com/ImageMagick/ImageMagick/assets/3990142/14ec9923-bb43-4add-8247-2ec3ada0e5ef)

**Comment by blue-j** [comment]
I think -format %[kurtosis] and the manual calculations are likely both correct but using different definitions, while the -fx method for these two metrics seems ahem askew.? - J

**Comment by snibgo** [comment]
> -0.0956531 -format %[kurtosis]

That's the result I get.

> -0.00000145957 -format %[fx:kurtosis]
> -0.000372191 -format %[fx:kurtosis*255]

I don't recommend either of these. Instead, as I suggested above, use:
```
C:\web\im>%IMG7%magick bigrock.jpg -format %[fx:kurtosis*QuantumRange] info:
-0.0956531
```

IM calculates kurtosis and skewness for each channel. It also calculates overall numbers, and I haven't tracked down how it does this.

Beware of overall numbers. Is the overall kurtosis set to the average kurtosis in each channel? Is that a valid thing to do? I don't know. For the overall kurtosis, IM doesn't find the average. Instead, it gives the same number as when we separate the channels into grayscale images and append them.
```
C:\web\im>%IMG7%magick bigrock.jpg -channel RGB -separate +append +repage -format %[fx:kurtosis*QuantumRange] info:
-0.0956531
```

Your "red rock" example, according to ImageJ, has a maximum value of 234 (out of 255). According to IM, the maximum is 255.
